src/hyperv/hyperv_network_api_v2.c

```diff
@@ -71,103 +71,141 @@ hyperv2MsvmVirtualSwitchToNetwork(virConnectPtr conn,
         goto cleanup;
 
-    entry = switches;
-    while (entry != NULL) {
-        if (VIR_STRDUP(names[count], entry->data->ElementName) < 0)
-            goto cleanup;
+    for (vSwitch = *list; vSwitch != NULL; vSwitch = vSwitch->next)
         count++;
-        if (count >= maxnames)
-            break;
-        entry = entry->next;
-    }
 
-    success = true;
+    ret = count;
 
-cleanup:
-    if (!success) {
-        for (i = 0; i < count; i++)
-            VIR_FREE(names[i]);
-
-        count = -1;
-    }
-    hypervFreeObject(priv, (hypervObject *) switches);
+ cleanup:
     virBufferFreeAndReset(&query);
-    return count;
+
+    return ret;
 }
 
+/* exported API functions */
+#define MATCH(FLAG) (flags & (FLAG))
 int
-hyperv2ConnectNumOfNetworks(virConnectPtr conn)
+hyperv2ConnectListAllNetworks(virConnectPtr conn,
+                              virNetworkPtr **networks,
+                              unsigned int flags)
 {
-    int result = -1, count = 0;
     hypervPrivate *priv = conn->privateData;
-    virBuffer query = VIR_BUFFER_INITIALIZER;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitchList = NULL;
-    Msvm_VirtualEthernetSwitch_V2 *virtualSwitch = NULL;
+    Msvm_VirtualEthernetSwitch_V2 *vSwitchList = NULL;
+    Msvm_VirtualEthernetSwitch_V2 *vSwitch = NULL;
+    virNetworkPtr network = NULL;
+    virNetworkPtr *nets = NULL;
+    int count = 0;
+    int ret = -1;
+    size_t nnets = 0;
+    size_t i = 0;
+
+    virCheckFlags(VIR_CONNECT_LIST_NETWORKS_FILTERS_ALL, -1);
+
+    /* filter out flag options that will produce 0 results:
+     * - inactive: all networks are active
+     * - transient: all networks are persistend
+     * - no autostart: all networks are auto start
+     */
+    if ((MATCH(VIR_CONNECT_LIST_NETWORKS_INACTIVE) &&
+         !MATCH(VIR_CONNECT_LIST_NETWORKS_ACTIVE)) ||
+        (MATCH(VIR_CONNECT_LIST_NETWORKS_TRANSIENT) &&
+         !MATCH(VIR_CONNECT_LIST_NETWORKS_PERSISTENT)) ||
+        (MATCH(VIR_CONNECT_LIST_DOMAINS_NO_AUTOSTART) &&
+         !MATCH(VIR_CONNECT_LIST_NETWORKS_AUTOSTART))) {
+        if (networks && VIR_ALLOC_N(*networks, 1) < 0)
+            goto cleanup;
 
-    virBufferAddLit(&query, MSVM_VIRTUALETHERNETSWITCH_V2_WQL_SELECT);
-    virBufferAsprintf(&query, " where HealthState = 5");
-    if (hyperv2GetMsvmVirtualEthernetSwitchList(priv, &query, &virtualSwitchList) < 0)
+        ret = 0;
         goto cleanup;
+    }
 
-    for (virtualSwitch = virtualSwitchList; virtualSwitch != NULL;
-         virtualSwitch = virtualSwitch->next) {
-        count++;
+    if (hyperv2GetVirtualSwitchList(priv, NULL, &vSwitchList) < 0)
+        goto cleanup;
+
+    if (networks) {
+        if (VIR_ALLOC_N(nets, 1) < 0)
+            goto cleanup;
+        nnets = 1;
     }
 
-    result = count;
+    for (vSwitch = vSwitchList; vSwitch != NULL; vSwitch = vSwitch->next) {
+        /* just count the networks */
+        if (!networks) {
```

Comment:
This block should probably be outside of the loop, since you know how many networks there are via `hyperv2GetVirtualSwitchList`.